In [1]:
import numpy as np

## Recursos de Numpy para deep learning
### Forma y dimesiones de matrices

In [3]:
#Obtener la forma y dimensión de una matriz

arr = np.array([[1, 2, 3, 4], [5, 6, 7, 8]])
print('La forma de arr es: ', arr.shape)
print('La matriz tiene ', arr.shape[0],"filas y ",arr.shape[1]," Columnas")
print('La dimensión de arr es: ', arr.ndim)
print('La matriz es: \n', arr)

La forma de arr es:  (2, 4)
La matriz tiene  2 filas y  4  Columnas
La dimensión de arr es:  2
La matriz es: 
 [[1 2 3 4]
 [5 6 7 8]]


In [4]:
#Para dar la dimensión deseada a una matriz

arr = np.array([1, 2, 3, 4], ndmin=5)
print('La matriz es: \n', arr)
print('La forma es:', arr.shape)
print('La dimensión es:', arr.ndim)

La matriz es: 
 [[[[[1 2 3 4]]]]]
La forma es: (1, 1, 1, 1, 4)
La dimensión es: 5


### Slice de matrices

In [10]:
#Tomar trozos de una matriz (slice)

a1 = arr[:,:,:,:,3]
a2 = arr[...,3]
a3 = arr[0][0][0][0][1:3]
a4 = arr[...,1:3]

print('Toma todos los valores de las dimensiones 0, 1, 2 y 3 \npero solo toma el indice 3 de la dimensión 4: \n', a1)
print('\nLo mismo se obtiene colocando elipses: \n', a2)
print('\nToma el valor de indice 0, de las dimensiones 0, 1, 2 y 3 \npero toma los valores entre el indice 1 a 3 de la dimensión 4: \n', a3)
print('\nLo mismo se obtiene colocando elipses: \n', a3)

Toma todos los valores de las dimensiones 0, 1, 2 y 3 
pero solo toma el indice 3 de la dimensión 4: 
 [[[[4]]]]

Lo mismo se obtiene colocando elipses: 
 [[[[4]]]]

Toma el valor de indice 0, de las dimensiones 0, 1, 2 y 3 
pero toma los valores entre el indice 1 a 3 de la dimensión 4: 
 [2 3]

Lo mismo se obtiene colocando elipses: 
 [2 3]


### Reshape de matrices

In [30]:
#Reformar las matrices

arr1 = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12])
print('arr1 es: \n', arr1)
print('La forma de arr1 es: ', arr1.shape)

newarr = arr1.reshape(4, 3) #Reforma con 1 submatriz
print("\nnewarr se reforma con reshape: \n", newarr)
print('La forma de newarr es: ', newarr.shape)

arr1 es: 
 [ 1  2  3  4  5  6  7  8  9 10 11 12]
La forma de arr1 es:  (12,)

newarr se reforma con reshape: 
 [[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]
La forma de newarr es:  (4, 3)


In [14]:
#Devuelve el objeto base si la memoria es de algún otro objeto.
print(newarr.base)

[ 1  2  3  4  5  6  7  8  9 10 11 12]


In [24]:
newarr1 = arr1.reshape(2, 3, 2) #Reforma con dos submatrices

print("\nLa matriz newarr1 se reforma con reshape \na dos submatrices de (3x2): \n", newarr1)
print('\nLa forma de newarr1 es: ', newarr1.shape)
print('\nLa forma de la matriz del indice 0 es: ', newarr1[0].shape)
print('\nLa submatriz del indice 0 es: \n', newarr1[0])


La matriz newarr1 se reforma con reshape 
a dos submatrices de (3x2): 
 [[[ 1  2]
  [ 3  4]
  [ 5  6]]

 [[ 7  8]
  [ 9 10]
  [11 12]]]

La forma de newarr1 es:  (2, 3, 2)

La forma de la matriz del indice 0 es:  (3, 2)

La submatriz del indice 0 es: 
 [[1 2]
 [3 4]
 [5 6]]


Se puede remodelar una matriz 1D de 8 elementos en una matriz 2D de 2 filas con 4 elementos cada una, pero no podemos reformarla en una matriz 2D de 3 filas con 3 elementos cada una, ya que eso requeriría 3x3 = 9 elementos. 

In [26]:
try:
    newarr = arr1.reshape(3, 3)
except ValueError as e:
    print('Este es el error: \n', e)
finally:
    print('\nEsta es la matriz: \n', arr1)

Este es el error: 
 cannot reshape array of size 12 into shape (3,3)

Esta es la matriz: 
 [ 1  2  3  4  5  6  7  8  9 10 11 12]


### Modificación de elementos

In [49]:
#Se crea una matriz
arr2 = np.array([1, 2, 3, 4, 5, 6, 7, 8])

#Se reforma a dos submatrices de (2,4)
#Se hace una copia
a = arr2.reshape(2,4)

print('La matriz de base es: \n', arr2)
print('\nLa forma de la matriz de base: \n', arr2.shape)
print('\nLa matriz reformada es: \n', a)
print('\nLa forma de la matriz reformada es: \n', a.shape)

La matriz de base es: 
 [1 2 3 4 5 6 7 8]

La forma de la matriz de base: 
 (8,)

La matriz reformada es: 
 [[1 2 3 4]
 [5 6 7 8]]

La forma de la matriz reformada es: 
 (2, 4)


In [50]:
#Modificamos el elemento de indice 0 de la submatriz de indice 0
a[0][0] = 45

print('\nLa matriz a con el elemento modificado es: \n', arr2)

#Verificamos la matriz original
print('La matriz de base es: \n', arr2)


La matriz a con el elemento modificado es: 
 [45  2  3  4  5  6  7  8]
La matriz de base es: 
 [45  2  3  4  5  6  7  8]


Se puede ver que la nueva matriz (a) lleva los valores como referencia, al modificarlos se modifican en la matriz original para evitar eso hay que hacer una *copia* de la matriz original.

In [51]:
#Se crea una matriz
arr2 = np.array([1, 2, 3, 4, 5, 6, 7, 8])

#Se reforma a dos submatrices de (2,4)
#Se hace una copia
a = arr2.reshape(2,4).copy()

#Modificamos el elemento de indice 0 de la submatriz de indice 0
a[0][0] = 45

print('La matriz de base es: \n', arr2)
print('\nLa forma de la matriz de base: \n', arr2.shape)
print('\nLa matriz reformada es: \n', a)
print('\nLa forma de la matriz reformada es: \n', a.shape)
print('\nLa matriz a con el elemento modificado es: \n', arr2)

La matriz de base es: 
 [1 2 3 4 5 6 7 8]

La forma de la matriz de base: 
 (8,)

La matriz reformada es: 
 [[45  2  3  4]
 [ 5  6  7  8]]

La forma de la matriz reformada es: 
 (2, 4)

La matriz a con el elemento modificado es: 
 [1 2 3 4 5 6 7 8]


### Dimensión desconocida

Se le permite tener una dimensión "*desconocida*". Esto se llama **matriz de rango 1**. Esto no es una matriz fila ni una matriz columna y puede generar errores en las operaciones matriciales.

In [62]:
#Matriz original con una dimensión desconocida
arr = np.array([1, 2, 3, 4, 5, 6, 7, 8])

#Reformamos la matriz para asignarle la dimensión desconocida
arr1 = arr.reshape(1,8)

print('La matriz original es: \n', arr)
print('Observar que tiene solo 1 corchete exterior')
print('\nLa forma de la matriz original es: \n', arr.shape)
print('Es una matriz de rango 1')
print('\nLa matriz reformada en la dimensión es: \n', arr1)
print('Observar que tiene 2 corchete exterior')
print('\nLa forma de la matriz original es: \n', arr1.shape)

La matriz original es: 
 [1 2 3 4 5 6 7 8]
Observar que tiene solo 1 corchete exterior

La forma de la matriz original es: 
 (8,)
Es una matriz de rango 1

La matriz reformada en la dimensión es: 
 [[1 2 3 4 5 6 7 8]]
Observar que tiene 2 corchete exterior

La forma de la matriz original es: 
 (1, 8)


In [67]:
#Para chequear la dimensión de la matriz
#Vemos que nos da un error, porque
#La forma de arr nos dio (8,1)

assert(arr.shape == (1,8))

AssertionError: 

In [68]:
#En el caso arr1 reformado, donde nos aseguramos la dimensión
#Nos va a verificar el assert
assert(arr1.shape == (1,8))

Por otra parte, es posible no tener que especificar un número exacto para una de las dimensiones en el método de remodelación.

In [63]:
#Reformamos la matriz sin indicar la dimensión
newarr = arr.reshape(2, 2, -1) #Se usa el comodín -1

print('\nLa matriz reformada es: \n', newarr)
print('\nLa forma de la matriz reformada es: \n', newarr.shape)


La matriz reformada es: 
 [[[1 2]
  [3 4]]

 [[5 6]
  [7 8]]]

La forma de la matriz reformada es: 
 (2, 2, 2)


### Transpuesta de una matriz

In [5]:
#Creamos una matriz
a = np.array([[1, 2, 3], [4, 5, 6]])

#Hacemos la transpuesta con T
b = a.T

#Hacemos la transpuesta con transpose
c = a.transpose()

print('La matriz original: \n', a)
print('\nLa matriz transpuesta con T: \n', b)
print('\nLa matriz transpuesta con transpose: \n', c)

La matriz original: 
 [[1 2 3]
 [4 5 6]]

La matriz transpuesta con T: 
 [[1 4]
 [2 5]
 [3 6]]

La matriz transpuesta con transpose: 
 [[1 4]
 [2 5]
 [3 6]]


En el siguiente ejemplo se ve porque es necesario definir las dimensiones antes de operar las matrices:

In [12]:
#Creamos una matriz aleatoria de 5 elementos
a = np.random.randn(3)

#Hacemos la Transpuesta
b = a.T

#Hacemos la multiplicación punto
c = np.dot(a, b)

print('La matriz original es: \n', a)
print('\nLa forma de la matriz es: \n', a.shape)
print('\nLa transpuesta de la matriz es: \n', b)
print('\nLa forma de la matriz transpuesta es: \n', b.shape)
print('\nEl resultado de la multiplicación es: \n', c)
print('\nLa forma de la multiplicación es: \n', c.shape)

La matriz original es: 
 [-0.35143978  1.29524288 -0.79627571]

La forma de la matriz es: 
 (3,)

La transpuesta de la matriz es: 
 [-0.35143978  1.29524288 -0.79627571]

La forma de la matriz transpuesta es: 
 (3,)

El resultado de la multiplicación es: 
 2.435219035583038

La forma de la multiplicación es: 
 ()


Como se observa, la matriz creada tiene forma (5,), que como se mencionó antes, es una matriz de rango 1 en python. Así, cuando se observa la transpuesta, acaba teniendo el mismo aspecto que la matriz original.

Si calculamos el producto punto de las dos matrices, se esperaría una matriz, pero devuelve un valor.

Lo que se recomienda es que cuando se programan redes neuronales, que no se usen estructuras de datos donde la forma sea (5,) o (n,). En su lugar, que a sea (5,1), entonces esto hace que "a" sea un vector columna de (5,1). Y mientras que previamente, "a" y "a transpuesta” parecían lo mismo, ahora a transpuesta llega a ser un vector fila.

In [11]:
#Creamos una matriz aleatoria de 5 elementos
a = np.random.randn(3,1) #Observe que se define filas y columna

#Hacemos la Transpuesta
b = a.T

#Hacemos la multiplicación punto
c = np.dot(a, b)

print('La matriz original es: \n', a)
print('\nLa forma de la matriz es: \n', a.shape)
print('\nLa transpuesta de la matriz es: \n', b)
print('\nLa forma de la matriz transpuesta es: \n', b.shape)
print('\nEl resultado de la multiplicación es: \n', c)
print('\nLa forma de la multiplicación es: \n', c.shape)

La matriz original es: 
 [[-0.43297044]
 [ 0.76470326]
 [ 0.01647248]]

La forma de la matriz es: 
 (3, 1)

La transpuesta de la matriz es: 
 [[-0.43297044  0.76470326  0.01647248]]

La forma de la matriz transpuesta es: 
 (1, 3)

El resultado de la multiplicación es: 
 [[ 1.87463404e-01 -3.31093908e-01 -7.13209792e-03]
 [-3.31093908e-01  5.84771073e-01  1.25965608e-02]
 [-7.13209792e-03  1.25965608e-02  2.71342671e-04]]

La forma de la multiplicación es: 
 (3, 3)


### Aplanar la matriz

In [2]:
#Creamos una matriz
arr3 = np.array([[1, 2, 3], [4, 5, 6]])

print('La matriz arr3 es: \n', arr3)
print('La forma de la matriz arr3 es: \n', arr3.shape)

La matriz arr3 es: 
 [[1 2 3]
 [4 5 6]]
La forma de la matriz arr3 es: 
 (2, 3)


In [73]:
#Aplanamos la matriz con reshape
newarr = arr3.reshape(-1) #Pero nos da con dimensión desconocida

print('La matriz newarr es: \n', newarr)
print('La forma de la matriz newarr es: \n', newarr.shape)

La matriz newarr es: 
 [1 2 3 4 5 6]
La forma de la matriz newarr es: 
 (6,)


In [75]:
#Aplanamos la matriz con reshape
#Pero que tenga la dimension definida
newarr = arr3.reshape(1,6)

print('La matriz newarr es: \n', newarr)
print('La forma de la matriz newarr es: \n', newarr.shape)

La matriz newarr es: 
 [[1 2 3 4 5 6]]
La forma de la matriz newarr es: 
 (1, 6)


In [78]:
#Aplanamos la matriz con flatten
arr4 = arr3.flatten()

print('La matriz arr4 es: \n', arr4)
print('La forma de la matriz arr4 es: \n', arr4.shape)

La matriz arr4 es: 
 [1 2 3 4 5 6]
La forma de la matriz arr4 es: 
 (6,)


In [79]:
#Aplanamos la matriz con ravel
arr5 = arr3.ravel()

print('La matriz arr5 es: \n', arr5)
print('La forma de la matriz arr5 es: \n', arr5.shape)

La matriz arr5 es: 
 [1 2 3 4 5 6]
La forma de la matriz arr5 es: 
 (6,)


**Diferencias entre Flatten() y Ravel():**

**a.ravel():**
* Devuelve solo la referencia / vista de la matriz original
* Si modifica la matriz, notará que el valor de la matriz original también cambia.
* Ravel es más rápido que flatten() ya que no ocupa memoria.
* Ravel es una función a nivel de biblioteca.

**a.flatten():**
* Devuelve una copia de la matriz original
* Si modifica cualquier valor de esta matriz, el valor de la matriz original no se ve afectado.
* Flatten() es comparativamente más lento que ravel() ya que ocupa memoria.
* Flatten es un método de un objeto ndarray.

## Vectorización
## Conceptos básicos

La **vectorización** se usa para acelerar el código Python sin usar loop. El uso de esta función puede ayudar a minimizar el tiempo de ejecución del código de manera eficiente.

En la era del deep learning, la seguridad en el aprendizaje profundo en la práctica, a menudo uno se encuentra entrenando conjuntos de datos relativamente grandes, porque ahí es cuando los algoritmos de aprendizaje profundo tienden a brillar. Así que es importante que el código sea muy rápido porque de lo contrario, se puede tardar mucho tiempo en ejecutar. Por esto es que la capacidad de realizar vectorización se ha convertido en una habilidad clave.

La complejidad del tiempo en la ejecución de cualquier algoritmo es crucial para decidir si una aplicación es confiable o no. Ejecutar un algoritmo grande en el mayor tiempo posible es muy importante cuando se trata de la aplicación de resultados en tiempo real. Para hacerlo, Python tiene algunas funciones matemáticas estándar para operaciones rápidas en matrices completas de datos sin tener que escribir bucles. Una de esas bibliotecas que contiene tal función es Numpy.

A continuación se presenta un ejemplo de vectorización usando `np.dot(w, x)`:

In [28]:
import time #Librería para medir tiempos
a = np.random.rand(1000000)    #Crea dos matrices con millones de valores aleatorios
b = np.random.rand(1000000) 

tic = time.time()    #Mide la hora actual
c = np.dot(a,b)      #Vectoriza la matriz
toc = time.time()    #Mide el tiempo en el proceso
print('Resultado de la multiplicación: ', c)
print('Versión Vectorizada: ' + str(1000*(toc-tic)) + ' ms')

c = 0
tic = time.time()    #Mide la hora actual
for i in range(1000000): #Bucle para hacer la multiplicación
    c += a[i]*b[i]
toc = time.time()    #Mide el tiempo en el proceso
print('\nResultado de la multiplicación: ', c)
print('Versión NO Vectorizada: ' + str(1000*(toc-tic)) + ' ms')

Resultado de la multiplicación:  249789.4103302722
Versión Vectorizada: 0.9987354278564453 ms

Resultado de la multiplicación:  249789.41033026483
Versión NO Vectorizada: 476.23181343078613 ms


En ambos casos, la versión vectorizada y la versión no vectorizada calculó los mismos valores, 249789. La versión vectorizada tomó alrededor de 1 milisegundos. La versión explícita para bucle y no vectorizada tomó cerca de 500 milisegundos. La versión no vectorizada tomó algo así como 300 veces más larga que la versión vectorizada.

Estás viendo cómo la vectorización puede acelerar significativamente tu código. **La regla de oro para recordar es siempre que sea posible, Evitar el uso de cuatro bucles explícitos**.

De hecho, la biblioteca NumPy tiene muchas otras funciones de valores vectoriales. Por ejemplo:
* `np.log(v)` calculará el logaritmo elemento a elemento. 
* `np.maximum(v,0)` calcula el valor máximo, puedes tomar el máximo de cada elemento de V con cero. 
* `V**2`, simplemente toma el cuadrado de cada elemento de V, 
* `1/V` toma el inverso de cada elemento y así sucesivamente. 

Entonces, cuando estés tentado a escribir un bucle for, echa un vistazo y ve si hay una manera de llamar una función incorporada de NumPy para hacerlo sin ese bucle for. 


## Broadcasting (transmisión)
### Conceptos iniciales

**Broadcasting** es una estrategia de numpy para operar con arreglos mutidimensionles de diferentes dimensiones.

El término Broadcasting describe cómo NumPy trata las matrices con diferentes formas durante las operaciones aritméticas. Sujeto a ciertas restricciones, el array más pequeño se "difunde" a través del array más grande para que tengan formas compatibles. 

El Broadcasting proporciona un medio de **vectorización de las operaciones de matrices** para que el bucle se produzca en C en lugar de Python. Esto lo hace sin hacer copias innecesarias de los datos y normalmente conduce a implementaciones de algoritmos eficientes.

Las operaciones de NumPy se realizan normalmente sobre pares de arrays *elemento a elemento*. En el caso más simple, los dos arrays deben tener exactamente la misma forma, como en el siguiente ejemplo:

In [6]:
# Dos array de mismas dimensiones
a = np.array([1.0, 2.0, 3.0])
b = np.array([2.0, 2.0, 2.0])
c = a * b
print('Multiplicación de arrays:', c)

Multiplicación de arrays: [2. 4. 6.]


La regla de Broadcasting de NumPy relaja esta restricción cuando las formas de las matrices cumplen con ciertas restricciones. El ejemplo de transmisión más simple ocurre cuando una matriz y un valor escalar se combinan en una operación:

In [8]:
a = np.array([1.0, 2.0, 3.0])  #Es un array
b = 2.0                        #Es un escalar
c = a * b                      #Puede ser cualquier operación
print('Multiplicación de arrays:', c)

Multiplicación de arrays: [2. 4. 6.]


Podemos pensar en el escalar **b** que se estira durante la operación aritmética en una matriz con la misma forma que **a**, como se muestra en la siguiente figura:

<figure>
 <img align="center", src="./imagenes/01.jpg", style="width:60%;" >
</figure>

En la siguiente figura se aprecia de manera similar cuando se estira a cada fila (mismo número de columnas):

<figure>
 <img align="center", src="./imagenes/02.jpg", style="width:60%;" >
</figure>

VENTAJA: El código del segundo ejemplo es más eficiente que el del primero porque el broadcasting mueve menos memoria durante la multiplicación (**b** es un escalar en lugar de una matriz).

### Reglas generales

Cuando opera en dos matrices, NumPy compara sus formas por elementos. Comienza con las dimensiones finales (es decir, más a la derecha) y avanza hacia la izquierda. 

Dos dimensiones son compatibles cuando
- son iguales, o
- uno de ellos es 1
- tienen misma cantidad de filas
- tienen misma cantidad de columnas

Si no se cumplen estas condiciones, se lanza una excepción, lo que indica que las matrices tienen formas incompatibles: *ValueError: operands could not be broadcast together*.

Ejemplos:

In [12]:
# Matriz con escalar ya lo vimos en el ejemplo anterior

# Dos matrices con misma forma de filas

a1 = np.array([[1.0, 2.0, 3.0], 
               [4.0, 5.0, 6.0]])
a2 = np.array([100.0, 200.0, 300.0])
c = a1 + a2 
print('Multiplicación de arrays:\n', c)

Multiplicación de arrays:
 [[101. 202. 303.]
 [104. 205. 306.]]


In [13]:
# Dos matrices con misma forma de columnas

a1 = np.array([[1.0, 2.0, 3.0], 
               [4.0, 5.0, 6.0]])
a2 = np.array([[100.0], 
               [200.0]])
c = a1 + a2 
print('Multiplicación de arrays:\n', c)

Multiplicación de arrays:
 [[101. 102. 103.]
 [204. 205. 206.]]


In [17]:
# Error de broadcasting

a1 = np.array([[1.0, 2.0], 
               [3.0, 4.0],
               [5.0, 6.0]]) #Dejo una columna menos que a2
a2 = np.array([100.0, 200.0, 300.0]) 
c = a1 + a2 
print('Multiplicación de arrays:\n', c)

ValueError: operands could not be broadcast together with shapes (3,2) (3,) 

In [18]:
# Error de broadcasting

a1 = np.array([[1.0, 2.0, 3.0], 
               [4.0, 5.0, 6.0]]) 
a2 = np.array([100.0, 200.0, 300.0, 400.0]) #Agrego una columna mas que a1
c = a1 + a2 
print('Multiplicación de arrays:\n', c)

ValueError: operands could not be broadcast together with shapes (2,3) (4,) 

Este último ejemplo se ilustra como:
<figure>
 <img align="center", src="./imagenes/03.jpg", style="width:50%;" >
</figure>

Las matrices no necesitan tener el mismo número de **dimensiones**. 

Por ejemplo, si tiene una matriz 256 x 256 x 3 (por ejemplo una imagen a color) y desea escalar cada color de la imagen por un valor diferente, puede multiplicar la imagen por una matriz unidimensional con 3 valores. La alineación de los tamaños de los ejes posteriores de estas matrices de acuerdo con las reglas de broadcastin muestra que son compatibles:

```
Image  (3d array): 256 x 256 x 3
Scale  (1d array):             3
Result (3d array): 256 x 256 x 3
```
Cuando alguna de las dimensiones comparadas es una, se utiliza la otra. En otras palabras, las dimensiones con tamaño 1 se estiran o "copian" para que coincidan con las otras.

En el siguiente ejemplo, tanto las matrices A como B tienen ejes con longitud uno que se expanden a un tamaño mayor durante la operación de transmisión:

```
A      (4d array):  8 x 1 x 6 x 1
B      (3d array):      7 x 1 x 5
Result (4d array):  8 x 7 x 6 x 5
```

El broadcasting proporciona una forma conveniente de tomar el producto externo (o cualquier otra operación externa) de dos arreglos. El siguiente ejemplo muestra una operación de suma externa de dos matrices unidimensionales:

In [20]:
A = np.array([0.0, 10.0, 20.0, 30.0])
B = np.array([1.0, 2.0, 3.0])
C = A[:, np.newaxis] + B

print('Forma de A: ', A.shape)
print('Forma de B: ', B.shape)
print('Resultado C: \n', C)
print('Forma de C: ', C.shape)

Forma de A:  (4,)
Forma de B:  (3,)
Resultado C: 
 [[ 1.  2.  3.]
 [11. 12. 13.]
 [21. 22. 23.]
 [31. 32. 33.]]
Forma de C:  (4, 3)


Gráficamente, el ejemplo anterior es:

<figure>
 <img align="center", src="./imagenes/04.jpg", style="width:50%;" >
</figure>

### Resumen

Para cualquier operación de matrices

| DIMENSIÓN MATRIZ A | DIMENSIÓN MATRIZ B | DIMENSIÓN FINAL |
| --- | --- | --- |
| (m, n) | (1, n) | (m, n) |
| (m, n) | (m, 1) | (m, n) |
| (m, 1) | (1, 1) | (m, 1) |
| (1, n) | (1, 1) | (1, n) |